### PR TITLE
EZP-27009: Configure the root of the UDW

### PR DESCRIPTION
**Pull request by StephaneDiot**
jira: https://jira.ez.no/browse/EZP-27009

## Description

Allow the UDW to start discovering only under a given root depth

## Tests

Unit and manually tested

## Screencast

https://youtu.be/-6Mxi56O4Ig

### CODE REVIEW

File: Resources/public/js/views/services/ez-universaldiscoveryviewservice.js [hunk 1]
```diff
@@ -98,6 +98,17 @@ YUI.add('ez-universaldiscoveryviewservice', function (Y) {
         ATTRS: {
 
             /**
+             * 
+             *
+             * @attribute discoverRootLocation
+             * @default false
+             * @type {eZ.Location|False}
+             */
+            discoverRootLocation: {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-universaldiscoveryviewservice.js]: seems useless :)

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js [hunk 2]
```diff
@@ -49,6 +49,11 @@ YUI.add('ez-universaldiscoveryfinderexplorerview', function (Y) {
                         this._addLevel(location);
                     }
                 }, this);
+                if (this.get('discoverRootDepth')) {
+                    for (var i = 0; i < this.get('discoverRootDepth'); i++) {
```
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js]: this is weird, this should be integrated in the loop above. That also means eZ.UniversalDiscoveryFinderExplorerLevelView should subscribed to `disabledChange` but it should expect the value to be set at build time.

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerlevelview.js [hunk 3]
```diff
@@ -327,6 +333,18 @@ YUI.add('ez-universaldiscoveryfinderexplorerlevelview', function (Y) {
...
              */
             childCount: {},
+
+            /**
+             * Boolean that tell if the view is disabled or not.
+             * The items of a disabled level view can not be explored.
+             * 
+             * @attribute disabled
+             * @default false
+             * @type Boolean
```
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerlevelview.js]: related to https://github.com/ezsystems/PlatformUIBundle/pull/814/commits/858da3cc40c35c5ca1e906d0834db2b09c45b9f5#r104382949, it should configured with `writeOnce: 'initOnly';`

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js [hunk 4]
```diff
@@ -214,6 +220,17 @@ YUI.add('ez-universaldiscoveryfinderexplorerview', function (Y) {
             },
 
             /**
+             * The discover root depth if the UDW is configured with one.
+             *
+             * @attribute discoverRootDepth
+             * @type {Number|false|Null}
```
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js]: that's a tad too much :) `{Number|false}` (with default to `false`) seems enough to me.

File: Resources/public/js/views/services/ez-discoverybarviewservice.js [hunk 5]
```diff
@@ -44,18 +44,21 @@ YUI.add('ez-discoverybarviewservice', function (Y) {
...
             }
 
             this.fire('contentDiscover', {
                 config: {
-                    title: 'Find location', //provisional
+                    title: Y.eZ.trans('universaldiscovery.content.browser', {}, 'universaldiscovery'),
                     multiple: false,
                     contentDiscoveredHandler: Y.bind(this._navigateToLocation, this),
                     startingLocationId: startingLocationId,
+                    discoverRootDepth: rootDepth,
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-discoverybarviewservice.js]: not super fan of the name `discoverRootDepth`, maybe `minDiscoverDepth` ?

File: Resources/public/js/views/ez-universaldiscoveryview.js [hunk 6]
```diff
@@ -518,6 +520,16 @@ YUI.add('ez-universaldiscoveryview', function (Y) {
             },
 
             /**
+             * The depth of the root where we start discovering content
```
Comment (dpobel) [inline Resources/public/js/views/ez-universaldiscoveryview.js]: this should explain that this parameter is related to the starting Location. Without a starting Location or with a value greater than the depth of the starting Location, this is useless.

File: Tests/js/views/services/assets/ez-discoverybarviewservice-tests.js [hunk 7]
```diff
@@ -149,13 +149,18 @@ YUI.add('ez-discoverybarviewservice-tests', function (Y) {
                     e.config.startingLocationId, this.locationId,
                     "startingLocationId should have the id of the current location of the locationViewView"
                 );
+
+                Assert.areSame(
+                    e.config.discoverRootDepth, 1,
+                    "discoverRootDepth be 1"
```
Comment (dpobel) [inline Tests/js/views/services/assets/ez-discoverybarviewservice-tests.js]: missing a word no ?

File: Tests/js/views/universaldiscovery/assets/ez-universaldiscoveryfinderexplorerlevelview-tests.js [hunk 8]
```diff
@@ -482,6 +483,44 @@ YUI.add('ez-universaldiscoveryfinderexplorerlevelview-tests', function (Y) {
         },
     });
 
+    disableTest = new Y.Test.Case({
```
Comment (dpobel) [inline Tests/js/views/universaldiscovery/assets/ez-universaldiscoveryfinderexplorerlevelview-tests.js]: should be `disabledTest` to be consistent with the tested attribute.

**Comment by StephaneDiot** [comment]
updated @dpobel

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerlevelview.js [hunk 9]
```diff
@@ -43,7 +44,7 @@ YUI.add('ez-universaldiscoveryfinderexplorerlevelview', function (Y) {
             this._fireMethod = this._fireLocationSearch;
             this._watchAttribute = 'items';
             container.addClass(IS_LOADING);
-
+
```
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerlevelview.js]: cs

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerlevelview.js (hunk 3 above)
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerlevelview.js]: you added the config for the attr but the doc is missing

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js [hunk 10]
```diff
@@ -41,12 +41,16 @@ YUI.add('ez-universaldiscoveryfinderexplorerview', function (Y) {
                 }
                 this._removeLevels();
                 this._getLevelViewPath().forEach(function (location, index, path) {
-                    var next = path[index + 1];
+                    var next = path[index + 1],
+                        disabled = false;
 
+                    if (this.get('minDiscoverDepth') > index) {
+                        disabled = true;
```
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js]: can be written in more elegant way:
```js
disabled = (this.get('minDiscoverDepth') > index);
```

File: Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js [hunk 11]
```diff
@@ -41,12 +41,16 @@ YUI.add('ez-universaldiscoveryfinderexplorerview', function (Y) {
...
+                    }
                     if ( next ) {
-                        this._addLevel(location, next.get('locationId'));
+                        this._addLevel(location, next.get('locationId'), disabled);
                     } else if ( location.get('childCount') !== 0 ) {
-                        this._addLevel(location);
+                        this._addLevel(location, null, disabled);
```
Comment (dpobel) [inline Resources/public/js/views/universaldiscovery/ez-universaldiscoveryfinderexplorerview.js]: I would pass `undefined` instead of null

**Comment by StephaneDiot** [comment]
updated @dpobel

**Comment by andrerom** [comment]
@dpobel Is this still intended for the release?

**Comment by dpobel** [comment]
@andrerom actually this is already in, @StephaneDiot forgot to clean up the intermediate PRs.

**Pull request closed without merging**